### DISCUSSION TITLE
Cilent send() method definition

### DISCUSSION THREAD

**Discussion by ZhaolinFang** [eclipse-emfcloud/emfcloud#206] [ModelServer] [answered]
Hi,

I encountered an issue when I try to use the client send() method(org.eclipse.emfcloud.modelserver.client.v2.ModelServerClientV2.send(String modelUri, String message)) to send a message to the server and broadcast it to other clients connected to the server. The error is: 

ERROR ContextResponse - Invalid JSON
com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'A': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
at [Source: (String)"A collaborator wants to stop the session, do you agree?"; line: 1, column: 2]
<img width="1261" alt="截屏2023-04-08 15 38 58" src="https://user-images.githubusercontent.com/90061850/230724346-612e8d19-7291-4bb4-87d3-fc53b104c2cc.png">.

I checked the definition of this method but I can not find what kind of message this method accepts, could you please provide one example of using this method?

**Comment by ndoschek** [comment, marked as answer]
Hi @ZhaolinFang,

A Model Server [Message](https://github.com/eclipse-emfcloud/emfcloud-modelserver/blob/master/bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/util/Message.java) basically consists of the message `type` and the message `data`. In the websocket use case, this message is typically a `string`.
Currently, there is only one type of message accepted by the Model Server, which is [the `keepAlive` message](https://github.com/eclipse-emfcloud/emfcloud-modelserver#websocket-endpoints).

This message would look like this:
```json
{ "type": "keepAlive", "data": "" }
```

If you would now try to send this message via the Java Client (which I assume from your description), this could look like this:

```java
JsonNode node = Json.object(
         Json.prop(JsonRequestMember.TYPE, Json.text("keepAlive")),
         Json.prop(JsonRequestMember.DATA, Json.text(""))
);
client.send(modelUri, node.toString());
```
or simpler:
```java
client.send(modelUri, "{\"type\":\"keepAlive\",\"data\":\"\"}");
```

As mentioned, the Model Server currently only accepts this type of websocket message.
If you would like to extend that functionality, a good starting point is the [DefaultSessionController#handleMessage](https://github.com/eclipse-emfcloud/emfcloud-modelserver/blob/master/bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultSessionController.java)

HTH and best regards,
Nina

**Reply by ZhaolinFang** [reply]
Thank you so much!

**Reply by ZhaolinFang** [reply]
Hi,

Sorry to ask again. I add a new message type according to the keepAlive and the message can be sent successfully, but I can not broadcast it to all clients that are connected to the same session. Could you please give me an example of it?

Specifically, We have several clients connected to the server on a model, and when one client wants to quit, he has to send a string message to tell other clients he will quit, and when the server receives this message, the server will broadcast this message to other clients.

Thank you in advance for your time.

Zhaolin

**Reply by ndoschek** [reply]
Hi @ZhaolinFang,

If you would like to handle a custom message type and broadcast it to all subscribed clients I can sketch a small for you example as follows:

For this example, my custom message's type would be "testString" and defined as `String TESTSTRING = "testString";` in `JsonResponseType`.

First, I would extend the `handleMessage` method in the `DefaultSessionController`
If I understand your use case correctly, you want to broadcast a message to all clients (independent of their modelUri) .
If a message with the custom type "testString" is received, immediately all subscribed client session get a notifcation.
```java
   @Override
   public boolean handleMessage(final WsMessageContext clientMessage) {
      if (!isClientSubscribed(clientMessage)) {
         return false;
      }
      if (isMessageType(clientMessage, JsonResponseType.KEEPALIVE)) {
         keepAlive(clientMessage);
         return true;
      }
      if (isMessageType(clientMessage, JsonResponseType.TESTSTRING)) {
         modelUrisToClientsV2.values().forEach(sessionSet -> sessionSet.forEach(session -> session.send(
            JsonResponse.customResponse("Broadcast in response to testString!"))));
         return true;
      }
      return false;
   }
```
In `JsonResponse` I would implement a custom response:
```java
   public static JsonNode customResponse(final String message) {
      return Json.merge(type("customResponse"), data(message));
   }
```

So to summarize:
A client would send: `ws.send(JSON.stringify({ type: 'testString', data: '' }))`
And all subscribed sessions receive via the websocket the following notifcation object `{"type":"customResponse","data":"Broadcast in response to testString!"}`

HTH and best regards,
Nina

**Reply by ZhaolinFang** [reply]
Thanks a lot for your patient answer.
